lib/resqued/quit-and-wait.rb

```diff
@@ -0,0 +1,83 @@
+require "optparse"
+
+module Resqued
+  class QuitAndWait
+    def self.exec!(argv)
+      options = {grace_seconds: 30}
+
+      opts = OptionParser.new do |opts|
+        opts.banner = <<USAGE
+Usage: resqued quit-and-wait PIDFILE [--grace-period SECONDS]
+
+Use this as a preStop script in kubernetes. This script will send a SIGQUIT to
+resqued immediately, and then sleep until either resqued exits or until the
+grace period is approximately 5 seconds from expiration. This script exits 0 if
```

Comment:
I wanted this to be a simple copy/paste from the container's `terminationGracePeriod`, but also to avoid having kubernetes kill this script off before it can produce its final log message. Kubernetes will wait `terminationGracePeriod` seconds for the pod to shut down cleanly, including this script. If it doesn't, it sends `SIGTERM` to everything. At this point, any running jobs will be interrupted but should have a chance to report an error. If the preStop script doesn't finish within the grace period, kube will add a small amount of time (2s) to the grace period for the same purpose, but I wanted to avoid relying on this. Also, like I said, I wanted to make sure that the preStop script is able to output its final message; this would let us use logs to differentiate between a broken script vs an unresponsive job.


See also kube's [container lifecycle hooks](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/) and [pod termination process](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination) docs.